polysemy-plugin/src/Polysemy/Plugin/InlineRecursiveCalls.hs

```diff
@@ -1,91 +1,153 @@
+import HsSyn
+
+
+------------------------------------------------------------------------------
+-- | Forces compiler to inline functions by creating loopbreaker with
+-- NO_INLINE pragma, changing recursive calls to use it and by adding INLINE
+-- pragma to the original function.
+inlineRecCallsAction :: (MonadUnique m, HasDynFlags m)
+                     => HsGroup GhcRn -> m (HsGroup GhcRn)
+inlineRecCallsAction group = do
+  dyn_flags <- getDynFlags
+
+  ( if optLevel dyn_flags > 0
+      then inlineRecCalls
+      else pure
+    ) group
+
+------------------------------------------------------------------------------
+inlineRecCalls :: MonadUnique m => HsGroup GhcRn -> m (HsGroup GhcRn)
+inlineRecCalls group@(hs_valds -> XValBindsLR (NValBinds binds sigs)) = do
+
+  (binds', extra_sigs) <- second concat . unzip
+                      <$> traverse inlineRecCall (pairTypes binds sigs)
+
+  pure group{ hs_valds = XValBindsLR $ NValBinds binds' $ sigs ++ extra_sigs }
+
+inlineRecCalls _ = error "inlineRecCalls: expected renamed group"
+
+------------------------------------------------------------------------------
+-- | Pairs single binds with their types if they have one specified.
+pairTypes :: [(RecFlag, LHsBinds GhcRn)]
+          -> [LSig GhcRn]
+          -> [((RecFlag, LHsBinds GhcRn), Maybe (LHsSigWcType GhcRn))]
+-- TODO: better way of doing this would be great
+pairTypes binds sigs = flip evalState sigs $ foldM go [] binds where
+  go pairs = \case
+    b@(_, bagToList -> [L _ FunBind{ fun_id = L _ fun_name }]) -> do
+      sig <- findMapPop $ \case
+        L _ (TypeSig _ sig_names bind_type)
+          | elem fun_name $ unLoc <$> sig_names -> Just bind_type
+        _                                       -> Nothing
+      pure $ (b, sig) : pairs
+
+    b -> pure $ (b, Nothing) : pairs
+
+------------------------------------------------------------------------------
+-- | Takes binding group and type of binding inside if there's only one and
+-- returns group with inlining of recursive calls and signatures to be added
+-- to the environment.
+inlineRecCall :: MonadUnique m
+              => ((RecFlag, LHsBinds GhcRn), Maybe (LHsSigWcType GhcRn))
+              -> m ((RecFlag, LHsBinds GhcRn), [LSig GhcRn])
+inlineRecCall (b, mtype)
+  | (Recursive, (bagToList -> [
+        L fun_loc fun_bind@FunBind{ fun_id = L _ fun_name, fun_matches }
+      ])) <- b
+  = do
+      (loopb_name, loopb_decl) <- loopbreaker fun_name
+
+      pure $ case maybeReplaceVarNames fun_name loopb_name fun_matches of
+        Just fun_matches' -> (,)
+          ( Recursive
+          , listToBag
+              [ L fun_loc fun_bind{ fun_matches = fun_matches' }
+              , loopb_decl
+              ]
+          )
+          (  [ inline alwaysInlinePragma                           fun_name
+             , inline defaultInlinePragma{ inl_inline = NoInline } loopb_name
+             ]
+          ++ maybeToList (loopbreakerSig loopb_name <$> mtype)
```

Comment:
What happens if this is `Nothing` --- we just get no sig? Does that work?

Reply:
If there was no sig to begin with, then it's probably infererable and so we shouldn't have to worry - plus we don't really have other options in renamer :smile:

Reply:
Fair enough! Maybe add a comment explaining that!

Reply:
Okay.